### PR TITLE
Fix virtual columns having incorrect values after ORDER BY

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed virtual columns (e.g. _file) showing incorrect values with ORDER BY.

Repro:
```
insert into function file('t1.csv') select 1 as x;
insert into function file('t2.csv') select 2 as x;
select _file, * from file('t{1,2}.csv') order by _file settings max_threads=1;

┌─_file──┬─c1─┐
│ t1.csv │  1 │
│ t1.csv │  2 │
└────────┴────┘
```

The storage was outputting `ColumnConst`, then sorting transform was concatenating them using `insertRangeFrom()`, which doesn't check if the values are equal.

I don't know what's the intended way for this to work.
 (1) Maybe storages are supposed to produce full columns only?
 (2) Maybe `ColumnConst::insertRangeFrom()` should throw an exception if values are different (this won't fix this problem but would make it more noticeable).
 (3) Maybe there should be something like `static IColumn::concat(ColumnPtr & a, IColumn & b)`, which would automatically convert `a` to full column if values are different?

This PR does the simplest thing (1) and just always materializes virtual columns. Lmk if that's not good enough.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 1c45326cea03f49e571a00a487899ff7468cd11e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54811/1c45326cea03f49e571a00a487899ff7468cd11e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

File: src/Storages/VirtualColumnUtils.cpp [hunk 1]
```diff
@@ -389,19 +389,19 @@ void addRequestedPathAndFileVirtualsToChunk(
     {
         if (virtual_column.name == "_path")
         {
-            chunk.addColumn(virtual_column.type->createColumnConst(chunk.getNumRows(), path));
+            chunk.addColumn(virtual_column.type->createColumnConst(chunk.getNumRows(), path)->convertToFullColumnIfConst());
```
Comment (nikitamikhaylov) [inline src/Storages/VirtualColumnUtils.cpp:392]: If we would agree to proceed with option (1), then the comment is needed which describes this behaviour + we need to check other places where we inject virtual columns.

**Comment by Avogar** [comment]
Originally virtual columns were materialized, but I changed it recently to ColumnConst to optimize queries like
```
select count(), _file from file(...) group by _file
```

With ColumnConst there are 2 optimizations:
1) Not materializing ColumnConst saves some time.
2) Recently I made an optimization for `group by` with constant keys.

We can compare master and build from this PR.
master:
```
avogar-dev :) select count(), _file, _path from file('data5.parquet') group by _file, _path

SELECT
    count(),
    _file,
    _path
FROM file('data5.parquet')
GROUP BY
    _file,
    _path

Query id: 9bf61dcc-5a3d-4ff8-b324-a72b74dd1a70

┌───count()─┬─_file─────────┬─_path───────────────────────────────┐
│ 100000000 │ data5.parquet │ /home/avogar/tmp/data/data5.parquet │
└───────────┴───────────────┴─────────────────────────────────────┘

1 row in set. Elapsed: 0.030 sec.
```

This PR:
```
avogar-dev :) select count(), _file, _path from file('data5.parquet') group by _file, _path

SELECT
    count(),
    _file,
    _path
FROM file('data5.parquet')
GROUP BY
    _file,
    _path

Query id: 1bca7eb3-5c53-4eae-a59f-ea4997f79a24

┌───count()─┬─_file─────────┬─_path───────────────────────────────┐
│ 100000000 │ data5.parquet │ /home/avogar/tmp/data/data5.parquet │
└───────────┴───────────────┴─────────────────────────────────────┘

1 row in set. Elapsed: 0.474 sec. Processed 86.80 million rows, 0.00 B (183.27 million rows/s., 0.00 B/s.)
Peak memory usage: 113.94 MiB.
```

Also we have such queries in Performance tests, but for some reason not all queries from `count_from_formats` perf test were executed.

So, I would prefer trying to make ColumnConst work with ORDER BY if it's possible

**Comment by Avogar** [comment]
Maybe we can merge this PR as fast fix (and backport it to 23.8) and then think of better solution

**Review by Avogar — APPROVED** [review]

**Comment by al13n321** [comment]
Trying to make ColumnConst work in https://github.com/ClickHouse/ClickHouse/pull/54866

**Comment by SaltTan** [comment]
@Avogar v23.8-must-backport ?

**Pull request merged**